diff -aurN weston-8.0.0/desktop-shell/shell.c weston-8.0.0-mod/desktop-shell/shell.c
--- weston-8.0.0/desktop-shell/shell.c	2021-07-30 07:09:10.530170154 -0700
+++ weston-8.0.0-mod/desktop-shell/shell.c	2021-07-30 07:11:17.825685366 -0700
@@ -2482,8 +2482,10 @@
 		pixman_region32_init(&surface->pending.input);
 		pixman_region32_fini(&surface->input);
 		pixman_region32_init(&surface->input);
+#if 0
 		weston_fade_run(shsurf->view, 1.0, 0.0, 300.0,
 				fade_out_done, shsurf);
+#endif
 	} else {
 		weston_view_destroy(shsurf->view);
 
