diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index c35c4ab34f07..9fbb9b187baa 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -540,14 +540,17 @@ void MediaPlayerPrivateGStreamerMSE::setRate(float rate)
     m_player->rateChanged();

     GstState state, pending;
+    GST_DEBUG(">>> AGPRINT1");
     gst_element_get_state(m_pipeline.get(), &state, &pending, 0);
     if (!rate && lastPlaybackRate == 1.0f) {
+        GST_DEBUG(">>> AGPRINT2");
         m_playbackRatePause = true;
         if (state != GST_STATE_PAUSED && pending != GST_STATE_PAUSED)
             changePipelineState(GST_STATE_PAUSED);
         return;
     }
     if (!lastPlaybackRate) {
+        GST_DEBUG(">>> AGPRINT3");
         m_playbackRatePause = false;
         if (state != GST_STATE_PLAYING && pending != GST_STATE_PLAYING)
             changePipelineState(GST_STATE_PLAYING);
